### ISSUE TITLE
what are query, keys and values in attention model?

### ISSUE THREAD

**Issue by lkluo** [kevinzakka/research-paper-notes#1]
Can you explain specifically, for example, in encoder layer?

**Comment by kevinzakka** [comment]
Let's start with an explanation of the attention mechanism in the regular **seq2seq** framework that you can easily extend to the self-attention model once you understand it.

<img width="812" alt="screen shot 2017-06-25 at 1 00 00 pm" src="https://user-images.githubusercontent.com/10518920/27519342-3da23692-59a6-11e7-88a3-39c07437df95.png">


The decoder implements an attention mechanism _over the hidden states of the encoder._ This enables it to "pay attention" to parts of the input, rather than relying on the last encoder hidden state (i.e. the *context* vector in orange) to carry the burden of encoding the entire meaning of the input sequence. Hence, for every step of the translation, the decoder selects a different part of the input sequence to consider and carries out the translation.

Concretely, this *attention* is calculated as follows. First, a set of attention weights are calculated using a function of the current encoder hidden states and the previous decoder hidden states. Think of this function as measuring how much the query (i.e. previous decoder hidden state) interacts with the input (i.e. encoder hidden state). These weights are usually normalized with a softmax function which squashes them to a `[0, 1]` range with a sum of 1.

Then, we compute a sum of these attention weights multiplied by the encoder hidden states to create a new context vector which feeds to the decoder RNN layer to create the outputs for the next time step. This weighted sum attenuates the parts of the input sequence that are not relevant to the current translation (the weights to these are close to 0) so that the decoder can **focus** and **attend** to the important parts.

<img width="628" alt="screen shot 2017-06-25 at 1 00 53 pm" src="https://user-images.githubusercontent.com/10518920/27519346-59471aac-59a6-11e7-9b0a-b0d70812570e.png">

**Note**: In the figure above, the decoder inputs are none other than the current encoder hidden states.

Check out Sean Robertson's [excellent tutorial](https://github.com/spro/practical-pytorch/blob/master/seq2seq-translation/seq2seq-translation.ipynb) on seq2seq with attention if you want a concrete implementation of attention (the images are taken from his tutorial).

**Comment by amin-3x** [comment]
The queries are the decoder hidden state in the previous time step. The keys are the encoder hidden states and I'm not sure about the values but I (think..) they're the same as the keys?

**Comment by gauravkoradiya** [comment]
@amin-3x you are right.
